libcaf_core/caf/flow/op/sample.hpp

```diff
@@ -0,0 +1,273 @@
+
+  sample_sub(coordinator* parent, observer<output_type> out)
+    : parent_(parent), out_(std::move(out)) {
+    // nop
+  }
+
+  // -- properties -------------------------------------------------------------
+
+  coordinator* parent() const noexcept override {
+    return parent_;
+  }
+
+  bool running() const noexcept {
+    return state_ == state::running;
+  }
+
+  const error& err() const noexcept {
+    return err_;
+  }
+
+  bool pending() const noexcept {
+    return buf_.has_value();
+  }
+
+  bool can_emit() const noexcept {
+    return buf_.has_value();
+  }
+
+  // -- callbacks for the parent -----------------------------------------------
+
+  void init(observable<input_type> vals, observable<select_token_type> ctrl) {
+    using val_fwd_t = forwarder<input_type, sample_sub, sample_input_t>;
+    using ctrl_fwd_t = forwarder<select_token_type, sample_sub, sample_emit_t>;
+    auto fwd = parent_->add_child(std::in_place_type<val_fwd_t>, this,
+                                  sample_input_t{});
+    vals.subscribe(fwd->as_observer());
+    // Note: the previous subscribe might call on_error, in which case we don't
+    // need to try to subscribe to the control observable.
+    if (running())
+      ctrl.subscribe(parent_->add_child_hdl(std::in_place_type<ctrl_fwd_t>,
+                                            this, sample_emit_t{}));
+  }
+
+  // -- callbacks for the forwarders -------------------------------------------
+
+  void fwd_on_subscribe(sample_input_t, subscription sub) {
+    if (!running() || value_sub_ || !out_) {
+      sub.cancel();
+      return;
+    }
+    value_sub_ = std::move(sub);
+    value_sub_.request(1);
+  }
+
+  void fwd_on_complete(sample_input_t) {
+    value_sub_.release_later();
+  }
+
+  void fwd_on_error(sample_input_t, const error& what) {
+    value_sub_.release_later();
+    err_ = what;
+  }
+
+  void fwd_on_next(sample_input_t, const input_type& item) {
+    if (running()) {
+      buf_ = item;
+      value_sub_.request(1);
+    }
+  }
+
+  void fwd_on_subscribe(sample_emit_t, subscription sub) {
+    if (!running() || control_sub_ || !out_) {
+      sub.cancel();
+      return;
+    }
+    control_sub_ = std::move(sub);
+    control_sub_.request(1);
+  }
+
+  void fwd_on_complete(sample_emit_t) {
+    control_sub_.release_later();
+    if (state_ == state::running)
+      err_ = make_error(sec::end_of_stream,
+                        "sample: unexpected end of the control stream");
+    shutdown();
+  }
+
+  void fwd_on_error(sample_emit_t, const error& what) {
+    control_sub_.release_later();
+    err_ = what;
+    shutdown();
+  }
+
+  void fwd_on_next(sample_emit_t, select_token_type) {
+    if (!value_sub_) {
+      shutdown();
+      return;
+    }
+    if (demand_ == 0)
+      return;
+    --demand_;
+    auto sampled = buf_.has_value();
+    if (sampled) {
+      out_.on_next(*buf_);
+      buf_.reset();
+    }
+    if (sampled)
+      value_sub_.request(1);
```

Comment:
This will give infinite credit to the source over time. We always call `value_sub_.request(1)` whenever receiving an item. Giving credit here as well will accumulate over time.

```suggestion
    if (buf_) {
      out_.on_next(*buf_);
      buf_.reset();
    }
```